File: web/src/apps/flappy.ts
import type { NeoPixel, RGB } from "../hardware/neopixel";
import type { Joystick, Pin } from "../hardware/joystick";
import * as screens from "../screens";
import { sleep_ms } from "../runtime/time";

export const NAME = "FlappyPixels";

const NUM_LEDS = 64;

const PLAYER_COL = 0;
const WALL_COLOR: RGB = [0, 40, 10];
const PLAYER_COLOR_GRAVITY: RGB = [60, 45, 0];
const PLAYER_COLOR_FLOAT: RGB = [0, 45, 55];

const FRAME_MS = 50;
const GAP_SIZE = 3;
const WALL_TICK = 6;
const SPAWN_TICK = 24;

const GRAVITY = 0.05;
const JUMP_DELTA = 0.4;
const JUMP_CAP = -1.0;
const TERMINAL_VEL = 1.5;

const FLOAT_ACCEL = 0.025;
const FLOAT_TERMINAL = 0.4;

type Mode = "gravity" | "float";
interface Wall { col: number; gap: number; }

let np: NeoPixel;
let JOY_UP: Pin, JOY_DOWN: Pin;
let JOY_SLIDE: Pin | undefined;

function slideMode(): Mode {
  if (!JOY_SLIDE) return "gravity";
  return JOY_SLIDE.value() === 0 ? "gravity" : "float";
}

function clear(): void {
  for (let i = 0; i < NUM_LEDS; i++) np[i] = [0, 0, 0];
}

function px(col: number, row: number, color: RGB): void {
  if (col >= 0 && col <= 7 && row >= 0 && row <= 7) {
    np[row * 8 + col] = color;
  }
}

function render(playerY: number, walls: ReadonlyArray<Wall>, mode: Mode): void {
  clear();
  for (const w of walls) {
    if (w.col < 0 || w.col > 7) continue;
    for (let r = 0; r < 8; r++) {
      const y = 7 - r;
      if (y < w.gap || y >= w.gap + GAP_SIZE) px(w.col, r, WALL_COLOR);
    }
  }
  const pyInt = Math.max(0, Math.min(7, Math.floor(playerY)));
  const color = mode === "float" ? PLAYER_COLOR_FLOAT : PLAYER_COLOR_GRAVITY;
  px(PLAYER_COL, 7 - pyInt, color);
  np.write();
}

function randInt(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

async function playOneGame(): Promise<number | null> {
  let playerY = 3.5;
  let velocity = 0.0;
  let walls: Wall[] = [];
  let score = 0;
  let frame = 0;
  let prevUp = false;
  const introFrames = 14;

  while (true) {
    if (screens.check_exit()) return null;

    const mode = slideMode();
    const curUp = JOY_UP.value() === 0;
    const curDown = JOY_DOWN.value() === 0;

    if (mode === "gravity") {
      if (curUp && !prevUp) velocity = Math.max(velocity - JUMP_DELTA, JUMP_CAP);
      velocity = Math.min(velocity + GRAVITY, TERMINAL_VEL);
    } else {
      if (curUp && !curDown) velocity -= FLOAT_ACCEL;
      else if (curDown && !curUp) velocity += FLOAT_ACCEL;
      if (velocity > FLOAT_TERMINAL) velocity = FLOAT_TERMINAL;
      else if (velocity < -FLOAT_TERMINAL) velocity = -FLOAT_TERMINAL;
    }

    prevUp = curUp;
    playerY += velocity;

    if (playerY < 0) {
      playerY = 0;
      if (velocity < 0) velocity = 0;
    }

    let groundHit = false;
    if (mode === "gravity") {
      groundHit = playerY >= 7.5;
    } else {
      if (playerY > 7) {
        playerY = 7;
        if (velocity > 0) velocity = 0;
      }
    }

    if (frame >= introFrames) {
      if (frame % WALL_TICK === 0) {
        const newWalls: Wall[] = [];
        let hit = false;
        for (const w of walls) {
          const newCol = w.col - 1;
          if (newCol === PLAYER_COL) {
            const pyInt = Math.max(0, Math.min(7, Math.floor(playerY)));
            if (w.gap <= pyInt && pyInt < w.gap + GAP_SIZE) {
              score += 1;
            } else {
              newWalls.push({ col: newCol, gap: w.gap });
              hit = true;
            }
          } else if (newCol >= 0) {
            newWalls.push({ col: newCol, gap: w.gap });
          }
        }
        walls = newWalls;
        if (hit) {
          render(playerY, walls, mode);
          return score;
        }
      }

      if ((frame - introFrames) % SPAWN_TICK === 0) {
        walls.push({ col: 7, gap: randInt(0, 7 - GAP_SIZE) });
      }
    }

    render(playerY, walls, mode);

    if (groundHit) return score;

    await sleep_ms(FRAME_MS);
    frame += 1;
  }
}

export async function run(neopixel: NeoPixel, joystick: Joystick): Promise<void> {
  np = neopixel;
  JOY_UP = joystick.up;
  JOY_DOWN = joystick.down;
  JOY_SLIDE = joystick.slide;
  screens.init(neopixel, joystick);
  while (true) {
    if ((await screens.loading_screen()) === "exit") return;
    const score = await playOneGame();
    if (score === null) return;
    if ((await screens.game_over_screen(score)) === "exit") return;
  }
}

File: python/apps/flappy.py
from machine import Pin
from neopixel import NeoPixel
from time import sleep_ms, ticks_ms, ticks_diff
import random
import _screens as screens

NAME = "FlappyPixels"
NUM_LEDS = 64

np = None
JOY_UP = None
JOY_DOWN = None
JOY_SLIDE = None

PLAYER_COL = 0
WALL_COLOR = (0, 40, 10)
PLAYER_COLOR_GRAVITY = (60, 45, 0)
PLAYER_COLOR_FLOAT   = (0, 45, 55)

FRAME_MS = 50
GAP_SIZE = 3
WALL_TICK = 6
SPAWN_TICK = 24

GRAVITY = 0.05
JUMP_DELTA = 0.40
JUMP_CAP = -1.0
TERMINAL_VEL = 1.5

FLOAT_ACCEL = 0.025
FLOAT_TERMINAL = 0.40


def slide_mode():
    if JOY_SLIDE is None:
        return "gravity"
    return "gravity" if JOY_SLIDE.value() == 0 else "float"


def clear():
    for i in range(NUM_LEDS):
        np[i] = (0, 0, 0)


def px(col, row, color):
    if 0 <= col <= 7 and 0 <= row <= 7:
        np[row * 8 + col] = color


def render(player_y, walls, mode):
    clear()
    for w in walls:
        col = w["col"]
        if col < 0 or col > 7:
            continue
        gap = w["gap"]
        for r in range(8):
            y = 7 - r
            if y < gap or y >= gap + GAP_SIZE:
                px(col, r, WALL_COLOR)
    py_int = max(0, min(7, int(player_y)))
    color = PLAYER_COLOR_FLOAT if mode == "float" else PLAYER_COLOR_GRAVITY
    px(PLAYER_COL, 7 - py_int, color)
    np.write()


def play_one_game():
    """Returns final score, or None if exit triggered mid-play."""
    player_y = 3.5
    velocity = 0.0
    walls = []
    score = 0
    frame = 0
    prev_up = False
    intro_frames = 14

    while True:
        if screens.check_exit():
            return None

        mode = slide_mode()
        cur_up = JOY_UP.value() == 0
        cur_down = JOY_DOWN.value() == 0

        if mode == "gravity":
            if cur_up and not prev_up:
                velocity = max(velocity - JUMP_DELTA, JUMP_CAP)
            velocity = min(velocity + GRAVITY, TERMINAL_VEL)
        else:
            if cur_up and not cur_down:
                velocity -= FLOAT_ACCEL
            elif cur_down and not cur_up:
                velocity += FLOAT_ACCEL
            if velocity > FLOAT_TERMINAL:
                velocity = FLOAT_TERMINAL
            elif velocity < -FLOAT_TERMINAL:
                velocity = -FLOAT_TERMINAL

        prev_up = cur_up
        player_y += velocity

        if player_y < 0:
            player_y = 0
            if velocity < 0:
                velocity = 0

        if mode == "gravity":
            ground_hit = player_y >= 7.5
        else:
            if player_y > 7:
                player_y = 7
                if velocity > 0:
                    velocity = 0
            ground_hit = False

        if frame >= intro_frames:
            if frame % WALL_TICK == 0:
                new_walls = []
                hit = False
                for w in walls:
                    new_col = w["col"] - 1
                    if new_col == PLAYER_COL:
                        py_int = max(0, min(7, int(player_y)))
                        gap = w["gap"]
                        if gap <= py_int < gap + GAP_SIZE:
                            score += 1
                        else:
                            new_walls.append({"col": new_col, "gap": gap})
                            hit = True
                    elif new_col >= 0:
                        new_walls.append({"col": new_col, "gap": w["gap"]})
                walls = new_walls
                if hit:
                    render(player_y, walls, mode)
                    return score

            if (frame - intro_frames) % SPAWN_TICK == 0:
                walls.append({"col": 7, "gap": random.randint(0, 7 - GAP_SIZE)})

        render(player_y, walls, mode)

        if ground_hit:
            return score

        sleep_ms(FRAME_MS)
        frame += 1


def run(neopixel, joystick):
    global np, JOY_UP, JOY_DOWN, JOY_SLIDE
    np = neopixel
    JOY_UP = joystick["up"]
    JOY_DOWN = joystick["down"]
    JOY_SLIDE = joystick.get("slide")
    screens.init(neopixel, joystick)
    while True:
        if screens.loading_screen() == "exit":
            return
        score = play_one_game()
        if score is None:
            return
        if screens.game_over_screen(score) == "exit":
            return


if __name__ == "__main__":
    _np = NeoPixel(Pin(19, Pin.OUT), NUM_LEDS)
    _joy = {
        "up":     Pin(3, Pin.IN),
        "down":   Pin(6, Pin.IN),
        "left":   Pin(7, Pin.IN),
        "right":  Pin(2, Pin.IN),
        "center": Pin(8, Pin.IN),
        "slide":  Pin(9, Pin.IN),
    }
    run(_np, _joy)
